File: packages/emitter/testcases/types/conditional/ConditionalTypes.ts
export type StringOrNumber = string | number | boolean;

// Using built-in conditional utility types
export type OnlyStrings = Extract<StringOrNumber, string>;
export type NoStrings = Exclude<StringOrNumber, string>;
export type NonNullableValue = NonNullable<string | null | undefined>;

// Custom conditional type
export type IsArray<T> = T extends any[] ? true : false;

export type ArrayCheck1 = IsArray<string[]>;
export type ArrayCheck2 = IsArray<string>;

// More complex conditional type
export type Unwrap<T> = T extends Promise<infer U> ? U : T;

export type UnwrappedString = Unwrap<Promise<string>>;
export type UnwrappedNumber = Unwrap<number>;

// Function using conditional type
export function processValue<T extends string | number>(
  value: T
): T extends string ? number : string {
  if (typeof value === "string") {
    return value.length as any;
  }
  return value.toString() as any;
}

// ReturnType utility
export function greet(name: string): string {
  return `Hello ${name}`;
}

export type GreetReturn = ReturnType<typeof greet>;

File: packages/emitter/testcases/types/conditional/ConditionalTypes.cs
using Tsonic.Runtime;
using Tsonic.JSRuntime;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace TestCases.types.conditional
{
        public static class ConditionalTypes
        {
            // type StringOrNumber = Union<string, double, bool>

            // type OnlyStrings = Extract<StringOrNumber, string>

            // type NoStrings = Exclude<StringOrNumber, string>

            // type NonNullableValue = NonNullable<string?>

            // type IsArray = dynamic

            // type ArrayCheck1 = IsArray<List<string>>

            // type ArrayCheck2 = IsArray<string>

            // type Unwrap = dynamic

            // type UnwrappedString = Unwrap<Task<string>>

            // type UnwrappedNumber = Unwrap<double>

            public static dynamic processValue<T>(T value)
                where T : Union<string, double>
                {
                if (Tsonic.Runtime.Operators.@typeof(value) == "string")
                    {
                    return value.length;
                    }
                return Tsonic.JSRuntime.Number.toString(value);
                }

            public static string greet(string name)
                {
                return $"Hello {name}";
                }

            // type GreetReturn = ReturnType<dynamic>
        }
}